front/components/assistant/conversation/AgentMessage.tsx

```diff
@@ -348,47 +351,68 @@ export function AgentMessage({
-            variant="ghost-secondary"
-            size="xs"
-            onClick={handleCopyToClipboard}
-            icon={isCopied ? ClipboardCheckIcon : ClipboardIcon}
-            className="text-muted-foreground"
-          />,
-          <Button
-            key="retry-msg-button"
-            tooltip="Retry"
-            variant="ghost-secondary"
-            size="xs"
-            onClick={() => {
-              void retryHandler({
-                conversationId,
-                messageId: agentMessageToRender.sId,
-              });
-            }}
-            icon={ArrowPathIcon}
-            className="text-muted-foreground"
-            disabled={isRetryHandlerProcessing || shouldStream}
-          />,
-          // One cannot leave feedback on global agents.
-          ...(isGlobalAgent ||
-          agentMessageToRender.configuration.status === "draft"
-            ? []
-            : [
-                <Separator key="separator" orientation="vertical" />,
-                <FeedbackSelector
-                  key="feedback-selector"
-                  {...messageFeedback}
-                  getPopoverInfo={PopoverContent}
-                />,
-              ]),
-        ];
+  const buttons: React.ReactElement[] = [];
+
+  // Standard buttons (visible when not thinking and not failed)
+  if (
+    message.status !== "failed" &&
+    messageStreamState.agentState !== "thinking"
+  ) {
+    buttons.push(
+      <Button
+        key="copy-msg-button"
+        tooltip={isCopied ? "Copied!" : "Copy to clipboard"}
+        variant="ghost-secondary"
+        size="xs"
+        onClick={handleCopyToClipboard}
+        icon={isCopied ? ClipboardCheckIcon : ClipboardIcon}
+        className="text-muted-foreground"
+      />,
+      <Button
+        key="retry-msg-button"
+        tooltip="Retry"
+        variant="ghost-secondary"
+        size="xs"
+        onClick={() => {
+          void retryHandler({
+            conversationId,
+            messageId: agentMessageToRender.sId,
+          });
+        }}
+        icon={ArrowPathIcon}
+        className="text-muted-foreground"
+        disabled={isRetryHandlerProcessing || shouldStream}
+      />,
+      // One cannot leave feedback on global agents.
+      ...(isGlobalAgent || agentMessageToRender.configuration.status === "draft"
+        ? []
+        : [
+            <Separator key="separator" orientation="vertical" />,
+            <FeedbackSelector
+              key="feedback-selector"
+              {...messageFeedback}
+              getPopoverInfo={PopoverContent}
+            />,
+          ])
+    );
+  }
+
+  // Add a per-agent stop control while generating.
+  if (agentMessageToRender.status === "created") {
+    buttons.push(
+      <Button
+        key="stop-msg-button"
+        tooltip="Stop agent"
+        variant="ghost-secondary"
+        size="xs"
+        onClick={async () => {
+          await cancelMessage([message.sId]);
+        }}
+        icon={XMarkIcon}
```

Comment:
`nit`: Don't we want to use the same square icon as for the "Stop all agents" button? As a matter of consistency